File: streamsql/modules/streaming_query/logical_plan.py
from __future__ import annotations

from dataclasses import dataclass, field
from enum import Enum
from typing import Any, Optional

from streamsql.core.models import generate_id


class NodeType(str, Enum):
    SCAN = "scan"
    FILTER = "filter"
    PROJECT = "project"
    JOIN = "join"
    AGGREGATE = "aggregate"
    WINDOW = "window"
    SORT = "sort"
    LIMIT = "limit"
    UNION = "union"
    INSERT = "insert"
    UPDATE = "update"
    DELETE = "delete"
    STREAM = "stream"
    WATERMARK = "watermark"


class JoinType(str, Enum):
    INNER = "inner"
    LEFT = "left"
    RIGHT = "right"
    FULL = "full"
    CROSS = "cross"
    LEFT_OUTER = "left_outer"
    RIGHT_OUTER = "right_outer"


class AggregateFunction(str, Enum):
    SUM = "SUM"
    AVG = "AVG"
    COUNT = "COUNT"
    MIN = "MIN"
    MAX = "MAX"
    COUNT_DISTINCT = "COUNT_DISTINCT"


@dataclass
class LogicalNode:
    node_id: str = field(default_factory=lambda: generate_id("ln"))
    node_type: NodeType = NodeType.SCAN
    children: list["LogicalNode"] = field(default_factory=list)
    properties: dict[str, Any] = field(default_factory=dict)
    estimated_cost: float = 0.0
    estimated_rows: Optional[int] = None

    def add_child(self, child: "LogicalNode") -> None:
        self.children.append(child)

    def get_all_nodes(self) -> list["LogicalNode"]:
        nodes = [self]
        for child in self.children:
            nodes.extend(child.get_all_nodes())
        return nodes

    def to_dict(self) -> dict[str, Any]:
        return {
            "node_id": self.node_id,
            "node_type": self.node_type.value,
            "children": [c.to_dict() for c in self.children],
            "properties": self.properties,
            "estimated_cost": self.estimated_cost,
            "estimated_rows": self.estimated_rows,
        }


@dataclass
class LogicalPlan:
    plan_id: str = field(default_factory=lambda: generate_id("lp"))
    root: Optional[LogicalNode] = None
    source_queries: list[str] = field(default_factory=list)
    optimized: bool = False
    metadata: dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> dict[str, Any]:
        return {
            "plan_id": self.plan_id,
            "root": self.root.to_dict() if self.root else None,
            "source_queries": self.source_queries,
            "optimized": self.optimized,
            "metadata": self.metadata,
        }

    def walk(self) -> list[LogicalNode]:
        if not self.root:
            return []
        return self.root.get_all_nodes()


class LogicalPlanner:
    def __init__(self):
        self.plan_id = generate_id("lp")

    def plan(self, parsed_query: Any) -> LogicalPlan:
        plan = LogicalPlan(source_queries=[parsed_query.raw_sql])

        if parsed_query.query_type.value == "SELECT":
            root = self._build_select_plan(parsed_query)
            plan.root = root
        elif parsed_query.query_type.value == "INSERT":
            plan.root = self._build_insert_plan(parsed_query)
        elif parsed_query.query_type.value == "UPDATE":
            plan.root = self._build_update_plan(parsed_query)
        elif parsed_query.query_type.value == "DELETE":
            plan.root = self._build_delete_plan(parsed_query)

        plan.metadata["tables"] = parsed_query.tables
        plan.metadata["is_streaming"] = parsed_query.is_streaming

        return plan

    def _build_select_plan(self, parsed_query: Any) -> LogicalNode:
        current: Optional[LogicalNode] = None

        for table in parsed_query.tables:
            scan_node = LogicalNode(
                node_type=NodeType.SCAN,
                properties={"table": table, "columns": [c["name"] for c in parsed_query.columns]},
            )
            if current is None:
                current = scan_node
            else:
                join_node = LogicalNode(
                    node_type=NodeType.JOIN,
                    properties={"join_type": JoinType.INNER.value},
                    children=[current, scan_node],
                )
                current = join_node

        if parsed_query.where_clause:
            filter_node = LogicalNode(
                node_type=NodeType.FILTER,
                properties={"condition": parsed_query.where_clause},
                children=[current] if current else [],
            )
            current = filter_node

        if parsed_query.group_by:
            agg_exprs = [c for c in parsed_query.columns if self._is_aggregate(c["expression"])]
            agg_node = LogicalNode(
                node_type=NodeType.AGGREGATE,
                properties={
                    "group_by": parsed_query.group_by,
                    "aggregations": agg_exprs,
                },
                children=[current] if current else [],
            )
            current = agg_node

        if parsed_query.window_spec:
            window_node = LogicalNode(
                node_type=NodeType.WINDOW,
                properties=parsed_query.window_spec,
                children=[current] if current else [],
            )
            current = window_node

        if parsed_query.order_by:
            sort_node = LogicalNode(
                node_type=NodeType.SORT,
                properties={"order_by": parsed_query.order_by},
                children=[current] if current else [],
            )
            current = sort_node

        if parsed_query.limit is not None:
            limit_node = LogicalNode(
                node_type=NodeType.LIMIT,
                properties={"limit": parsed_query.limit, "offset": parsed_query.offset or 0},
                children=[current] if current else [],
            )
            current = limit_node

        project_node = LogicalNode(
            node_type=NodeType.PROJECT,
            properties={"columns": parsed_query.columns},
            children=[current] if current else [],
        )
        current = project_node

        if parsed_query.is_streaming:
            stream_node = LogicalNode(
                node_type=NodeType.STREAM,
                properties={},
                children=[current] if current else [],
            )
            current = stream_node

        return current or LogicalNode(node_type=NodeType.SCAN, properties={})

    def _is_aggregate(self, expr: str) -> bool:
        agg_funcs = ["SUM(", "AVG(", "COUNT(", "MIN(", "MAX("]
        return any(f in expr.upper() for f in agg_funcs)

    def _build_insert_plan(self, parsed_query: Any) -> LogicalNode:
        return LogicalNode(
            node_type=NodeType.INSERT,
            properties={"table": parsed_query.tables[0] if parsed_query.tables else "unknown"},
        )

    def _build_update_plan(self, parsed_query: Any) -> LogicalNode:
        return LogicalNode(
            node_type=NodeType.UPDATE,
            properties={
                "table": parsed_query.tables[0] if parsed_query.tables else "unknown",
                "where": parsed_query.where_clause,
            },
        )

    def _build_delete_plan(self, parsed_query: Any) -> LogicalNode:
        return LogicalNode(
            node_type=NodeType.DELETE,
            properties={
                "table": parsed_query.tables[0] if parsed_query.tables else "unknown",
                "where": parsed_query.where_clause,
            },
        )

    def estimate_cost(self, plan: LogicalPlan) -> float:
        if not plan.root:
            return 0.0
        return self._estimate_node_cost(plan.root)

    def _estimate_node_cost(self, node: LogicalNode) -> float:
        cost = 0.0
        for child in node.children:
            cost += self._estimate_node_cost(child)

        base_costs = {
            NodeType.SCAN: 10.0,
            NodeType.FILTER: 5.0,
            NodeType.PROJECT: 2.0,
            NodeType.JOIN: 100.0,
            NodeType.AGGREGATE: 50.0,
            NodeType.WINDOW: 75.0,
            NodeType.SORT: 200.0,
            NodeType.LIMIT: 1.0,
            NodeType.STREAM: 25.0,
        }
        cost += base_costs.get(node.node_type, 10.0)

        node.estimated_cost = cost
        return cost

File: src/streaming_sql/logical_plan.rs
use serde::{Deserialize, Serialize};
use std::collections::HashSet;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LogicalPlan {
    pub plan_id: String,
    pub root: LogicalNode,
    pub estimated_cost: f64,
    pub created_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum LogicalNode {
    Scan(ScanNode),
    Filter(FilterNode),
    Project(ProjectNode),
    Join(JoinNode),
    Aggregate(AggregateNode),
    Window(WindowNode),
    Watermark(WatermarkNode),
    Union(UnionNode),
    Sort(SortNode),
    Limit(LimitNode),
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScanNode {
    pub table_name: String,
    pub columns: Vec<String>,
    pub is_streaming: bool,
    pub predicate: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FilterNode {
    pub condition: String,
    pub child: Box<LogicalNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectNode {
    pub columns: Vec<ProjectExpression>,
    pub child: Box<LogicalNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectExpression {
    pub name: String,
    pub expression: String,
    pub alias: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JoinNode {
    pub join_type: JoinType,
    pub left: Box<LogicalNode>,
    pub right: Box<LogicalNode>,
    pub condition: Option<String>,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum JoinType {
    Inner,
    Left,
    Right,
    Full,
    Cross,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AggregateNode {
    pub group_by: Vec<String>,
    pub aggregations: Vec<AggregationExpression>,
    pub child: Box<LogicalNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AggregationExpression {
    pub function: AggregationFunction,
    pub column: Option<String>,
    pub alias: String,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum AggregationFunction {
    Count,
    Sum,
    Avg,
    Min,
    Max,
    First,
    Last,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WindowNode {
    pub window_type: super::WindowType,
    pub duration_ms: u64,
    pub slide_duration_ms: Option<u64>,
    pub time_column: String,
    pub child: Box<LogicalNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct WatermarkNode {
    pub column: String,
    pub delay_ms: u64,
    pub child: Box<LogicalNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UnionNode {
    pub left: Box<LogicalNode>,
    pub right: Box<LogicalNode>,
    pub all: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SortNode {
    pub columns: Vec<SortColumn>,
    pub child: Box<LogicalNode>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SortColumn {
    pub name: String,
    pub ascending: bool,
    pub nulls_first: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LimitNode {
    pub limit: u64,
    pub offset: u64,
    pub child: Box<LogicalNode>,
}

pub struct LogicalPlanBuilder {
    plan_id: String,
}

impl LogicalPlanBuilder {
    pub fn new() -> Self {
        Self {
            plan_id: format!("lp_{}", uuid::Uuid::new_v4()),
        }
    }

    pub fn from_parsed_query(&self, parsed: &super::ParsedQuery) -> LogicalPlan {
        let mut root = self.build_scan_nodes(parsed);

        if parsed.is_streaming {
            if let Some(watermark) = &parsed.watermark {
                root = LogicalNode::Watermark(WatermarkNode {
                    column: watermark.column.clone(),
                    delay_ms: watermark.delay_ms,
                    child: Box::new(root),
                });
            }

            if let Some(window) = &parsed.window_spec {
                root = LogicalNode::Window(WindowNode {
                    window_type: window.window_type,
                    duration_ms: window.duration_ms,
                    slide_duration_ms: window.slide_duration_ms,
                    time_column: window.time_column.clone(),
                    child: Box::new(root),
                });
            }
        }

        LogicalPlan {
            plan_id: self.plan_id.clone(),
            root,
            estimated_cost: self.estimate_cost(&root),
            created_at: chrono::Utc::now(),
        }
    }

    fn build_scan_nodes(&self, parsed: &super::ParsedQuery) -> LogicalNode {
        if parsed.source_tables.is_empty() {
            LogicalNode::Scan(ScanNode {
                table_name: "dual".to_string(),
                columns: vec!["1".to_string()],
                is_streaming: parsed.is_streaming,
                predicate: None,
            })
        } else if parsed.source_tables.len() == 1 {
            LogicalNode::Scan(ScanNode {
                table_name: parsed.source_tables[0].clone(),
                columns: vec!["*".to_string()],
                is_streaming: parsed.is_streaming,
                predicate: None,
            })
        } else {
            let first = LogicalNode::Scan(ScanNode {
                table_name: parsed.source_tables[0].clone(),
                columns: vec!["*".to_string()],
                is_streaming: parsed.is_streaming,
                predicate: None,
            });

            parsed.source_tables[1..].iter().fold(first, |acc, table| {
                LogicalNode::Join(JoinNode {
                    join_type: JoinType::Inner,
                    left: Box::new(acc),
                    right: Box::new(LogicalNode::Scan(ScanNode {
                        table_name: table.clone(),
                        columns: vec!["*".to_string()],
                        is_streaming: parsed.is_streaming,
                        predicate: None,
                    })),
                    condition: None,
                })
            })
        }
    }

    fn estimate_cost(&self, node: &LogicalNode) -> f64 {
        match node {
            LogicalNode::Scan(scan) => {
                let base_cost = if scan.is_streaming { 100.0 } else { 10.0 };
                base_cost * (scan.columns.len() as f64)
            }
            LogicalNode::Filter(f) => 1.2 * self.estimate_cost(&f.child),
            LogicalNode::Project(p) => 1.1 * self.estimate_cost(&p.child),
            LogicalNode::Join(j) => {
                10.0 * (self.estimate_cost(&j.left) + self.estimate_cost(&j.right))
            }
            LogicalNode::Aggregate(a) => 5.0 * self.estimate_cost(&a.child),
            LogicalNode::Window(w) => 8.0 * self.estimate_cost(&w.child),
            LogicalNode::Watermark(w) => 1.05 * self.estimate_cost(&w.child),
            LogicalNode::Union(u) => self.estimate_cost(&u.left) + self.estimate_cost(&u.right),
            LogicalNode::Sort(s) => 2.0 * self.estimate_cost(&s.child),
            LogicalNode::Limit(l) => 0.5 * self.estimate_cost(&l.child),
        }
    }

    pub fn build(&self, root: LogicalNode) -> LogicalPlan {
        LogicalPlan {
            plan_id: self.plan_id.clone(),
            root,
            estimated_cost: self.estimate_cost(&root),
            created_at: chrono::Utc::now(),
        }
    }
}

impl Default for LogicalPlanBuilder {
    fn default() -> Self {
        Self::new()
    }
}

impl LogicalPlan {
    pub fn tables(&self) -> HashSet<String> {
        let mut tables = HashSet::new();
        self.collect_tables(&self.root, &mut tables);
        tables
    }

    fn collect_tables(&self, node: &LogicalNode, tables: &mut HashSet<String>) {
        match node {
            LogicalNode::Scan(scan) => {
                tables.insert(scan.table_name.clone());
            }
            LogicalNode::Filter(f) => self.collect_tables(&f.child, tables),
            LogicalNode::Project(p) => self.collect_tables(&p.child, tables),
            LogicalNode::Join(j) => {
                self.collect_tables(&j.left, tables);
                self.collect_tables(&j.right, tables);
            }
            LogicalNode::Aggregate(a) => self.collect_tables(&a.child, tables),
            LogicalNode::Window(w) => self.collect_tables(&w.child, tables),
            LogicalNode::Watermark(w) => self.collect_tables(&w.child, tables),
            LogicalNode::Union(u) => {
                self.collect_tables(&u.left, tables);
                self.collect_tables(&u.right, tables);
            }
            LogicalNode::Sort(s) => self.collect_tables(&s.child, tables),
            LogicalNode::Limit(l) => self.collect_tables(&l.child, tables),
        }
    }

    pub fn is_streaming(&self) -> bool {
        self.check_streaming(&self.root)
    }

    fn check_streaming(&self, node: &LogicalNode) -> bool {
        match node {
            LogicalNode::Scan(scan) => scan.is_streaming,
            LogicalNode::Filter(f) => self.check_streaming(&f.child),
            LogicalNode::Project(p) => self.check_streaming(&p.child),
            LogicalNode::Join(j) => self.check_streaming(&j.left) || self.check_streaming(&j.right),
            LogicalNode::Aggregate(a) => self.check_streaming(&a.child),
            LogicalNode::Window(_) => true,
            LogicalNode::Watermark(_) => true,
            LogicalNode::Union(u) => self.check_streaming(&u.left) || self.check_streaming(&u.right),
            LogicalNode::Sort(s) => self.check_streaming(&s.child),
            LogicalNode::Limit(l) => self.check_streaming(&l.child),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_logical_plan_builder() {
        let builder = LogicalPlanBuilder::new();
        let scan = LogicalNode::Scan(ScanNode {
            table_name: "orders".to_string(),
            columns: vec!["*".to_string()],
            is_streaming: true,
            predicate: None,
        });

        let plan = builder.build(scan);
        assert!(plan.is_streaming());
        assert!(plan.tables().contains("orders"));
    }

    #[test]
    fn test_join_plan() {
        let builder = LogicalPlanBuilder::new();
        let left = LogicalNode::Scan(ScanNode {
            table_name: "orders".to_string(),
            columns: vec!["*".to_string()],
            is_streaming: true,
            predicate: None,
        });
        let right = LogicalNode::Scan(ScanNode {
            table_name: "users".to_string(),
            columns: vec!["*".to_string()],
            is_streaming: false,
            predicate: None,
        });

        let join = LogicalNode::Join(JoinNode {
            join_type: JoinType::Inner,
            left: Box::new(left),
            right: Box::new(right),
            condition: Some("orders.user_id = users.id".to_string()),
        });

        let plan = builder.build(join);
        assert!(plan.is_streaming());
        assert!(plan.tables().contains("orders"));
        assert!(plan.tables().contains("users"));
    }
}
